File: analysis/end_helpers.py
# Ethan Doughty
# end_helpers.py
"""Utilities for resolving the End keyword in indexing expressions."""

from __future__ import annotations
from typing import Optional

from ir import Expr, End, Const, BinOp


def _binop_contains_end(expr: Expr) -> bool:
    """Recursively check if End appears in a BinOp tree."""
    if isinstance(expr, End):
        return True
    if isinstance(expr, BinOp):
        return _binop_contains_end(expr.left) or _binop_contains_end(expr.right)
    return False


def _eval_end_arithmetic(expr: Expr, end_value: int) -> Optional[int]:
    """Evaluate a BinOp tree with End resolved to end_value.

    Returns None if can't resolve (e.g., contains variables).
    """
    if isinstance(expr, End):
        return end_value
    if isinstance(expr, Const):
        return int(expr.value)
    if isinstance(expr, BinOp):
        left = _eval_end_arithmetic(expr.left, end_value)
        right = _eval_end_arithmetic(expr.right, end_value)
        if left is None or right is None:
            return None
        if expr.op == '+':
            return left + right
        if expr.op == '-':
            return left - right
        if expr.op == '*':
            return left * right
        if expr.op == '/':
            return left // right if right != 0 else None
        # Unsupported operator
        return None
    return None  # Can't resolve (e.g., Var in expression)

File: src/EndHelpers.fs
module EndHelpers

open Ir

// ---------------------------------------------------------------------------
// Utilities for resolving the End keyword in indexing expressions.
// Port of analysis/end_helpers.py
// ---------------------------------------------------------------------------

/// binopContainsEnd: recursively check if End appears in a BinOp tree.
let rec binopContainsEnd (expr: Expr) : bool =
    match expr with
    | End _ -> true
    | BinOp(_, _, _, left, right) -> binopContainsEnd left || binopContainsEnd right
    | _ -> false


/// evalEndArithmetic: evaluate a BinOp tree with End resolved to endValue.
/// Returns None if can't resolve (e.g., contains variables).
let rec evalEndArithmetic (expr: Expr) (endValue: int) : int option =
    match expr with
    | End _ -> Some endValue
    | Const(_, _, v) -> Some (int v)
    | BinOp(_, _, op, left, right) ->
        match evalEndArithmetic left endValue, evalEndArithmetic right endValue with
        | Some l, Some r ->
            match op with
            | "+" -> Some (l + r)
            | "-" -> Some (l - r)
            | "*" -> Some (l * r)
            | "/" -> if r <> 0 then Some (l / r) else None
            | _   -> None
        | _ -> None
    | _ -> None   // Var or other nodes: can't resolve
